chain-evm/src/machine.rs

```diff
@@ -12,64 +12,245 @@
+    }
+
+    /// Creates a new `VirtualMachine` given configuration params and a given account storage.
+    pub fn new_with_state(
+        config: &'runtime Config,
+        environment: &'runtime Environment,
+        state: AccountTrie,
+    ) -> Self {
         Self {
+            config,
             environment,
+            precompiles: precompiles(),
             state,
             logs: Default::default(),
         }
     }
 
-    #[allow(dead_code)]
-    /// Returns an initialized instance of `evm::executor::StackExecutor`.
-    fn executor<'backend, 'config>(
-        &'backend self,
-        gas_limit: u64,
-        config: &'config Config,
-    ) -> StackExecutor<'config, MemoryStackState<'backend, 'config, VirtualMachine>> {
-        let metadata = StackSubstateMetadata::new(gas_limit, config);
-        let memory_stack_state = MemoryStackState::new(metadata, self);
-        StackExecutor::new(memory_stack_state, config)
-    }
-
-    #[allow(dead_code)]
     /// Returns an initialized instance of `evm::Runtime`.
-    fn runtime<'config>(
+    pub fn runtime(
         &self,
         code: Rc<Vec<u8>>,
         data: Rc<Vec<u8>>,
         context: RuntimeContext,
-        config: &'config Config,
-    ) -> Runtime<'config> {
-        Runtime::new(code, data, context, config)
+    ) -> Runtime<'_> {
+        Runtime::new(code, data, context, self.config)
+    }
+
+    /// Execute a CREATE transaction
+    #[allow(clippy::boxed_local)]
+    pub fn transact_create(
+        &mut self,
+        caller: Address,
+        value: Value,
+        init_code: ByteCode,
+        gas_limit: u64,
+        access_list: Vec<(Address, Vec<Key>)>,
+        delete_empty: bool,
+    ) -> Option<(&AccountTrie, &[Log])> {
+        {
+            let metadata = StackSubstateMetadata::new(gas_limit, self.config);
+            let memory_stack_state = MemoryStackState::new(metadata, self);
+            let mut executor = StackExecutor::new_with_precompiles(
+                memory_stack_state,
+                self.config,
+                &self.precompiles,
+            );
+
+            let exit_reason =
+                executor.transact_create(caller, value, init_code.to_vec(), gas_limit, access_list);
+            match exit_reason {
+                ExitReason::Succeed(_succeded) => {
+                    // apply and return state
+                    // apply changes to the state, this consumes the executor
+                    let state = executor.into_state();
+                    // Next, we consume the stack state and extract the values and logs
+                    // used to modify the accounts trie in the VirtualMachine.
+                    let (values, logs) = state.deconstruct();
+
+                    self.apply(values, logs, delete_empty);
+                    //_exit_reason
+                    Some((&self.state, &self.logs))
+                }
+                _ => None,
```

Comment:
Could it be useful for debugging to return the error here?

Reply:
Yes, the other variants for `ExitReason` need to be properly handled. It should be straightforward to convert into ledger errors or logs. The reason that this method returns an `Option<_>` is to keep things consistent with how the EVM crate works.